src/replay/service.zig

```diff
@@ -31,128 +36,471 @@ const MAX_ENTRIES: u64 = 1024 * 1024; // 1 million slots is about 5 days
+        var votes_per_slot = self.votes_per_slot;
+        for (votes_per_slot.values()) |*htvm| {
+            for (htvm.values()) |*vl| vl.deinit(allocator);
+            htvm.deinit(allocator);
+        }
+        votes_per_slot.deinit(allocator);
+    }
+
+    /// Update `latest_validator_votes_for_frozen_slots` if gossip has seen a newer vote for a frozen slot.
+    pub fn addVote(
+        self: *UnfrozenGossipVerifiedVoteHashes,
+        allocator: std.mem.Allocator,
+        vote_pubkey: Pubkey,
+        vote_slot: Slot,
+        hash: Hash,
+        is_frozen: bool,
+        latest_validator_votes_for_frozen_slots: *LatestValidatorVotes,
+    ) !void {
+        // If this is a frozen slot, then we need to update the `latest_validator_votes_for_frozen_slots`
+        const was_added, //
+        const maybe_latest_frozen_vote_slot //
+        = if (is_frozen) try latest_validator_votes_for_frozen_slots.checkAddVote(
+            allocator,
+            vote_pubkey,
+            vote_slot,
+            hash, // is_frozen
+            .gossip,
+        ) else blk: {
+            // Non-frozen banks are not inserted because
+            // we only track frozen votes in this struct
+            const vote_map = latest_validator_votes_for_frozen_slots.latestVotes(.gossip);
+            const slot = if (vote_map.get(vote_pubkey)) |entry| entry.slot else null;
+            break :blk .{ false, slot };
+        };
+
+        const vote_slot_gt_latest_frozen_slot: bool = blk: {
+            const latest_frozen_vote_slot = maybe_latest_frozen_vote_slot orelse {
+                // If there's no latest frozen vote slot yet, then we should also insert
+                break :blk true;
+            };
+            break :blk vote_slot >= latest_frozen_vote_slot;
+        };
+        if (!was_added and vote_slot_gt_latest_frozen_slot) {
+            // At this point it must be that:
+            // 1) `vote_slot` was not yet frozen
+            // 2) and `vote_slot` >= than the latest frozen vote slot.
+
+            // Thus we want to record this vote for later, in case a slot with this `vote_slot` + hash gets
+            // frozen later
+            const vps_gop = try self.votes_per_slot.getOrPut(allocator, vote_slot);
+            errdefer if (!vps_gop.found_existing) {
+                std.debug.assert(self.votes_per_slot.orderedRemove(vps_gop.key_ptr.*));
+            };
+            const hash_to_votes: *HashToVotesMap = vps_gop.value_ptr;
+
+            if (!vps_gop.found_existing) {
+                hash_to_votes.* = .empty;
+            }
+
+            const htv_gop = try hash_to_votes.getOrPut(allocator, hash);
+            errdefer if (!htv_gop.found_existing) {
+                std.debug.assert(hash_to_votes.swapRemove(htv_gop.key_ptr.*));
+            };
+
+            try htv_gop.value_ptr.append(allocator, vote_pubkey);
+        }
+    }
 };
```

Comment:
Is any of this going to be used outside edge_cases.zig? I think anything that's exclusive to the edge cases code makes sense to live within edge_cases.zig, just like ConsensusDependencies and ReplayExecutionState live in their respective files.

Reply:
I moved this originally for some specific organizational limitation I encountered, but isn't a problem now after something I did.

Reply:
I'm not too sure either; I think it'll be easier to just have it live in service for now, and if it's still only used by edge_cases as a whole, then we can move it there - let practice define our decision.

Reply:
8b1109e2e853d4ec17ec272b9d52992503369023